### DISCUSSION TITLE
Reaction forces with sympy.physics.mechanics

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#27077] [Q&A]
Someone sent me this sketch:
![example-fig-924 (002)](https://github.com/user-attachments/assets/37aa8a87-3389-414b-a9ba-2adef55b467b)

When the disk accelerates, there should be reaction forces from the shaft on D and on E, provided the bearings at D and E cannot take torques.
It is easy to get the reaction forces on, say, D and the reaction torques, too. I did it with the little program attached.
But I could not come up with a way to describe the system in sympy mechanics so I can get the reaction forces on D **and** E.

_Am I missing something, or is it not possible in the frmework of sympy mechanics?_

_I have never used the Joints Framework, as I feel more comfortable with reference frames.
Is it possible to do things with Joints, which cannot, in principle be done with reference frames - even if much more clumsyly so?_

**Thanks for any hrlp!**

```
import sympy as sn
import sympy.physics.mechanics as me

# %%
theta = sn.symbols("theta")
phi, psi = me.dynamicsymbols("phi psi")
uphi, upsi = me.dynamicsymbols("uphi upsi")

t = me.dynamicsymbols._t

xd, yd, zd, uxd, uyd, uzd = me.dynamicsymbols('xd, yd, zd, uxd, uyd, uzd')
xR, yR, zR, uxR, uyR, uzR = me.dynamicsymbols('xR, yR, zR, uxR, uyR, uzR')

Td = me.dynamicsymbols("T_D")
# reaction forces / torques acting on point D, R
Dx, Dy, Dz, tRx, tRy, tRz = reactions = me.dynamicsymbols("D_x, D_y, D_z, tR_x, tR_y, tR_z")
# auxiliary velocities needed to compute the reaction forces /torques
u1_aux, u2_aux, u3_aux, u4_aux, u5_aux, u6_aux = u_aux = me.dynamicsymbols("u_aux^((1:7))")

# N: inertial frame
# M: frame of the rotating platform
# R: intermediate frame
# S: frame of the shaft
N, M, R, S = sn.symbols("N, M, R, S", cls=me.ReferenceFrame)
M.orient_axis(N, N.z, phi)
M.set_ang_vel(N, N.z * uphi)
R.orient_axis(M, M.x, theta)
R.set_ang_vel(M, u4_aux * R.x + u5_aux * R.y + u6_aux * R.z)
S.orient_axis(R, R.x, psi)
S.set_ang_vel(R, R.x *upsi)

# %%
b, h, l, m, g = sn.symbols("b, h, l, m, g")
O, G, D, E, Gh, Dh, Eh = sn.symbols("O, G, D, E, Gh, Dh, Eh", cls=me.Point)
G.set_pos(O, 2 * N.z)
Eh.set_pos(G, l/2*R.x)
Dh.set_pos(G, -l/2*R.x)

D.set_pos(O, xd*N.x + yd*N.y + zd*N.z)
D.set_vel(N, uxd*N.x + uyd *N.y + uzd*N.z + u1_aux * N.x + u2_aux * N.y + 
    u3_aux * N.z)
E.set_pos(D, l*R.x)
E.v2pt_theory(D, N, R)

Gh.set_pos(D, l/2*R.x)
O.set_vel(N, 0)

config_const_D = sn.Matrix([(D.pos_from(O) - Dh.pos_from(O)).dot(uv) 
        for uv in N])
vel_constr = config_const_D.diff(t)

# %%
q_ind = [phi, psi, xd, yd, zd]
u_ind = [uphi, upsi]
u_dep = [uxd, uyd, uzd]
kde = [uphi - phi.diff(t), upsi - psi.diff(t)] +  [i - j.diff(t) for i, j in 
        zip(u_dep, (xd, yd, zd))]

kane = me.KanesMethod(
    N,
    q_ind=q_ind,
    u_ind=u_ind,
    u_dependent = u_dep,
    kd_eqs=kde,
    u_auxiliary=u_aux,
    velocity_constraints=vel_constr,
)

A, B, C = sn.symbols("A B C")
central_inertia = me.inertia(S, A, B, C)
bodies = [
    me.RigidBody("shaft", Gh, S, m, (central_inertia, Gh))
]
loads = [
    (Gh, -m * g * N.z),
    (D, Dx * N.x + Dy * N.y + Dz * N.z),
    (R, tRx * N.x + tRy * N.y + tRz * N.z),
    (S, Td * S.x),
    (M, Td * N.z),
]
fr, frstar = kane.kanes_equations(bodies, loads)
eom = fr + frstar
print(me.find_dynamicsymbols(eom))

rhs = kane.rhs()

# %%
reaction_forces = kane.auxiliary_eqs
reaction_forces = sn.solve(reaction_forces, [Dx, Dy, Dz, tRx, tRy, tRz])

for i in (Dx, Dy, Dz, tRx, tRy, tRz):
    print(str(i) + ':')
    print(reaction_forces[i], '\n')
```

**Comment by moorepants** [comment]
The general answer is that any constraint force or torque can be extracted using the method of auxiliary speeds.

**Reply by Peter230655** [reply]
> The general answer is that any constraint force or torque can be extracted using the method of auxiliary speeds.

That's what I thought, too.
In the example, I have to go to point G (mass center of the shaft) either from point D or from point E.
If I take, say, E that I can easily get the reaction forces and torques there - but of course no forces on E.
(Of course, with the reaction torques at E I could calculate manually the forces at D if E was unable to take torques, but this is not what I mean)
**Thanks!**

**Reply by moorepants** [reply]
If D and E have auxiliary speeds added to them, then G will have a velocity that is a function of both the auxiliary speeds from D and E. My guess is that you have to carefully calculate the velocity of G so that all auxiliary speeds are present.

**Comment by Peter230655** [comment]
Clever!
So, I would still define G with reference to E (?)
The auxiliary sppeds of G are function (I guess linear ones here) of those at E and D.
There are no reaction forces at G (?)
I will try your idea tomorrow!

**Comment by Peter230655** [comment]
The reaction forces at E, D due to the mass of the body with mass center G work fine with your suggestion, but I do not get the dynamic reaction forces - at least not yet.
